src/consensus/network_event.rs

```diff
@@ -95,21 +99,104 @@ pub enum AccumulatingEvent {
+        match self {
+            Self::Genesis { .. } | Self::StartDkg(_) | Self::DkgResult { .. } => false,
+            _ => true,
+        }
     }
 
-    pub fn into_network_event(self) -> NetworkEvent {
-        NetworkEvent {
+    /// Sign and convert this event into `NetworkEvent`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.needs_signature()` is `false`
+    pub fn into_signed_network_event(
+        self,
+        section_key_share: &SectionKeyShare,
+    ) -> Result<NetworkEvent> {
+        let signature_share = self.sign(&section_key_share.secret_key_share)?;
+        let proof_share = ProofShare {
+            public_key_set: section_key_share.public_key_set.clone(),
+            index: section_key_share.index,
+            signature_share,
+        };
+
+        Ok(NetworkEvent {
             payload: self,
-            signature: None,
-        }
+            proof_share: Some(proof_share),
+        })
     }
 
-    pub fn into_network_event_with(self, signature: Option<bls::SignatureShare>) -> NetworkEvent {
+    /// Convert this event into unsigned `NetworkEvent`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.needs_signature()` is `true`
+    #[cfg(all(test, feature = "mock"))]
+    pub fn into_unsigned_network_event(self) -> NetworkEvent {
+        assert!(!self.needs_signature());
+
         NetworkEvent {
             payload: self,
-            signature,
+            proof_share: None,
+        }
+    }
+
+    /// Create a signature share of this event using `secret key share`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.needs_signature()` is `false`
+    pub fn sign(&self, secret_key_share: &bls::SecretKeyShare) -> Result<bls::SignatureShare> {
+        assert!(self.needs_signature());
+
+        let bytes = self.serialise_for_signing()?;
+        Ok(secret_key_share.sign(&bytes))
+    }
+
+    pub fn verify(&self, proof_share: &ProofShare) -> bool {
+        self.needs_signature()
+            && self
+                .serialise_for_signing()
+                .map(|bytes| proof_share.verify(&bytes))
+                .unwrap_or(false)
+    }
+
+    fn serialise_for_signing(&self) -> Result<Vec<u8>, bincode::Error> {
+        match self {
+            Self::OurKey { prefix: _, key } => {
+                // Note: the prefix is only needed to differentiate between the two subsections in
+                // case of split but it isn't part of the proven data.
+                bincode::serialize(key)
+            }
+            Self::TheirKey { prefix, key } => bincode::serialize(&(prefix, key)),
+            Self::TheirKnowledge { prefix, knowledge } => bincode::serialize(&(prefix, knowledge)),
+            Self::SectionInfo(info) => bincode::serialize(info),
+            Self::Online {
+                p2p_node,
+                age: _,
+                their_knowledge: _,
+            } => bincode::serialize(&member_info::to_sign(p2p_node.name(), MemberState::Joined)),
```

Comment:
knowledge doesn't because it is not stored. It is used only to create correct proof chain for the `NodeApproval` message. `age` should be signed eventually, but I decided to postpone doing that because there is a change to the node ageing and relocation logic planned and it would most likely affect the way we do this and I wanted to avoid spending time on something that's likely going to change soon anyway.